sessiontxn/isolation/base.go

```diff
@@ -175,14 +176,34 @@ func (p *baseTxnContextProvider) activateTxn() (kv.Transaction, error) {
 		return nil, err
 	}
 
-	txn, err := p.sctx.Txn(true)
+	txn, err := p.sctx.GetPreparedTxnFuture().Wait(p.ctx, p.sctx)
 	if err != nil {
 		return nil, err
 	}
 
 	sessVars := p.sctx.GetSessionVars()
 	sessVars.TxnCtx.StartTS = txn.StartTS()
 
+	if !sessVars.IsAutocommit() && sessVars.SnapshotTS == 0 {
+		sessVars.SetInTxn(true)
+	}
+
+	sessVars.TxnCtx.CouldRetry = sessiontxn.IsTxnRetryable(sessVars)
```

Comment:
TxnCtx.CouldRetry can only be true when it is a optimistic txn, so it's better to move set it in `OptimisticTxnContextProvider.onTxnActive`, for other providers, keep it always false.

Reply:
BTW, session.getSnapshotInterceptor can be removed